Cabal/Distribution/Backpack/ComponentsGraph.hs

```diff
@@ -0,0 +1,80 @@
+-- | See <https://github.com/ezyang/ghc-proposals/blob/backpack/proposals/0000-backpack.rst>
+
+module Distribution.Backpack.ComponentsGraph (
+    ComponentsGraph,
+    dispComponentsGraph,
+    toComponentsGraph,
+    componentCycleMsg
+) where
+
+import Distribution.Package
+import Distribution.PackageDescription as PD hiding (Flag)
+import Distribution.Simple.LocalBuildInfo
+import Distribution.Types.ComponentRequestedSpec
+import Distribution.Simple.Utils
+import Distribution.Compat.Graph (Node(..))
+import qualified Distribution.Compat.Graph as Graph
+
+import Distribution.Text
+    ( Text(disp) )
+import Text.PrettyPrint
+
+------------------------------------------------------------------------------
+-- Components graph
+------------------------------------------------------------------------------
+
+-- | A components graph is a source level graph tracking the
+-- dependencies between components in a package.
+type ComponentsGraph = [(Component, [ComponentName])]
+
+-- | Pretty-print a 'ComponentsGraph'.
+dispComponentsGraph :: ComponentsGraph -> Doc
+dispComponentsGraph graph =
+    vcat [ hang (text "component" <+> disp (componentName c)) 4
+                (vcat [ text "dependency" <+> disp cdep | cdep <- cdeps ])
+         | (c, cdeps) <- graph ]
+
+-- | Given the package description and the set of package names which
+-- are considered internal (the current package name and any internal
```

Comment:
> and the set of package names which are considered internal

This isn't correct anymore right? The old `mkComponentGraph` did take a map of internal component names, but this new impl doen't take any such thing.